sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReader.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Iterates over all rows in a table.
+ */
+class BigQueryParallelReader<T> extends BoundedSource.BoundedReader<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReader.class);
+
+  private Session session;
+  private final BigQueryServices client;
+  private Iterator<ReadRowsResponse> responseIterator;
+  private Iterator<Row> rowIterator;
+  private final ReadRowsRequest request;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private BoundedSource<T> source;
+  private BigQueryOptions options;
+  private Row currentRow;
+
+  private BigQueryParallelReader(
+      Session session,
+      ReadLocation readLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    this.session = checkNotNull(session, "session");
+    this.client = checkNotNull(client, "client");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.source = checkNotNull(source, "source");
+    this.options = options;
+
+    ReadRowsRequest.Builder builder = ReadRowsRequest.newBuilder().setReadLocation(readLocation);
+    if (rowBatchSize != null) {
+      builder.setOptions(ReadOptions.newBuilder().setMaxRows(rowBatchSize).build());
+    }
+
+    this.request = builder.build();
+  }
+
+  public static <T> BigQueryParallelReader<T> create(
+      Session session,
+      ReadLocation initialReadLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    return new BigQueryParallelReader<>(
+        session,
+        initialReadLocation,
+        rowBatchSize,
+        parseFn,
+        source,
+        client,
+        options);
+  }
+  /**
+   * Empty operation, the table is already open for read.
+   * @throws IOException on failure
+   */
+  @Override
+  public boolean start() throws IOException {
+    responseIterator = client.getTableReadService(options).readRows(request);
+    if (responseIterator != null && responseIterator.hasNext()) {
+      rowIterator = responseIterator.next().getRowsList().iterator();
+      if (rowIterator.hasNext()) {
+        currentRow = rowIterator.next();
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    if (responseIterator == null) {
+      throw new IOException("start() needs to be called first");
+    }
+    if (rowIterator != null && rowIterator.hasNext()) {
+      currentRow = rowIterator.next();
+      return true;
+    }
+    if (rowIterator == null || !rowIterator.hasNext()) {
+      if (responseIterator.hasNext()) {
+        rowIterator = responseIterator.next().getRowsList().iterator();
```

Comment:
Is it possible for the API to return an empty rows list in the current ReadRowsResponse, but return a non-empty one in the subsequent call? I.e. does this need to be wrapped in a loop that gives up only after `responseIterator` is exhausted?

Reply:
It is not possible, it always returns the expected row count or according to the byte limit. If a single row exceeds the byte limit, then the stream will error out (not ideal but it is the current behavior).